# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

createdb: database creation failed: ERROR:  database "pagila_star" already exists
 set_config 
------------
 
(1 row)

psql:Data/pagila-star.sql:42: ERROR:  type "mpaa_rating" already exists
psql:Data/pagila-star.sql:52: ERROR:  type "year" already exists
psql:Data/pagila-star.sql:69: ERROR:  function "_group_concat" already exists with same argument types
psql:Data/pagila-star.sql:86: ERROR:  function "film_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:103: ERROR:  function "film_not_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:148: ERROR:  function "get_customer_balance" already exists with same argument types
psql:Data/pagila-star.sql:170: ERROR:  function "inventory_held_by_customer" already exists with same argument types
psql:Data/pagila-star.sql:207: ERROR:  function "inventory_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:225: ERROR:  function "last_day" already exists with same argument 

 setval 
--------
  32098
(1 row)

psql:Data/pagila-star.sql:65348: ERROR:  duplicate key value violates unique constraint "rental_pkey"
DETAIL:  Key (rental_id)=(2) already exists.
CONTEXT:  COPY rental, line 1
 setval 
--------
  16049
(1 row)

psql:Data/pagila-star.sql:65365: ERROR:  duplicate key value violates unique constraint "staff_pkey"
DETAIL:  Key (staff_id)=(1) already exists.
CONTEXT:  COPY staff, line 1
 setval 
--------
      2
(1 row)

psql:Data/pagila-star.sql:65382: ERROR:  duplicate key value violates unique constraint "store_pkey"
DETAIL:  Key (store_id)=(1) already exists.
CONTEXT:  COPY store, line 1
 setval 
--------
      2
(1 row)

psql:Data/pagila-star.sql:65397: ERROR:  multiple primary keys for table "actor" are not allowed
psql:Data/pagila-star.sql:65405: ERROR:  multiple primary keys for table "address" are not allowed
psql:Data/pagila-star.sql:65413: ERROR:  multiple primary keys for table "category" are not allowed
psql:Data/pagila-star.sql:65421: ERROR:

### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [3]:
%%sql
SELECT sum(sales_amount) as total_revenue
FROM factSales;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.


[(Decimal('67416.51'),)]

In [13]:
# %%sql
# SELECT ...
# FROM ...

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [5]:
%%sql
SELECT country, sum(sales_amount) as total_revenue
FROM factSales f
join dimstore ds on f.store_key = ds.store_key
group by 1
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
2 rows affected.


[('Australia', Decimal('33726.77')), ('Canada', Decimal('33689.74'))]

In [14]:
# %%sql
# SELECT ...
# FROM ...

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [7]:
%%sql
SELECT month, sum(sales_amount) as total_revenue
FROM factSales f
join dimdate dd on f.date_key = dd.date_key
group by 1
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.


[(1, Decimal('4824.43')),
 (3, Decimal('23886.56')),
 (4, Decimal('28559.46')),
 (2, Decimal('9631.88')),
 (5, Decimal('514.18'))]

In [15]:
# %%sql
# SELECT ...
# FROM ...

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [9]:
%%sql
SELECT month, country, sum(sales_amount) as total_revenue
FROM factSales f
join dimdate dd on f.date_key = dd.date_key
join dimstore ds on f.store_key = ds.store_key
group by 1,2
order by 1,2,3 desc
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (3, 'Australia', Decimal('12060.33'))]

In [16]:
# %%sql
# SELECT ...
# FROM ...

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [11]:
%%sql
SELECT month, country, sum(sales_amount) as total_revenue
FROM factSales f
join dimdate dd on f.date_key = dd.date_key
join dimstore ds on f.store_key = ds.store_key
group by grouping sets ((), month,country, (month, country))

 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74'))]

In [17]:
# %%sql
# SELECT ...
# FROM ...

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>